requirements.txt

```diff
@@ -2,6 +2,7 @@ PyYAML==6.0
 attrs==21.4.0
 certifi==2021.10.8
 charset-normalizer==2.0.10
+coverage==6.2
```

Comment:
This package is not used any more with pytest's in-built coverage support right?

Reply:
It appears to be getting installed by a dependency chain. I haven't looked to see what specifically is doing it. `pytest` is getting installed here rather than in `-dev` the same way :-/

I have a half-baked requirements.txt management script that I plan to turn into something real in the future. It'll probably start with the setup.py `install_requires` and `tests_requires` and build fresh and up to date requirements.txt and requirements-dev.txt files off of that. 

That should keep what's in here minimal (based on the chains of our requirements at least) and up to date. I think that'll be ideal

Reply:
Ok cool yeah it'll be nice to not have to put transitive dependencies and keep this minimal at some point.

Reply:
The transitive nature is actually intentional. That ensures that dev & ci happen with known versions and an update to a downstream package doesn't cause CI failures or if someone checks out the repo and bootstraps they'll get to a known/fixed state that should be happy/passing. 

Without pinning everything, deps with `>=` would install the latest version on each bootstrap and potentially break things if there's bugs or incompatibilities in the updates, or worse if someone intentionally pushes a broken version as has happened a couple times in the past few years.

My rough plan is to call out what's explicit and implicit in the requirements files, something like:

```txt
# Explicit requirements
dnspython=2.1.0
requests=0.27.0
# Transitive requirements
attrs==21.4.0
certifi==2021.10.8
charset-normalizer==2.0.10
coverage==6.2
...
```

Reply:
If I was pinning a package version, I'd generally expect its dependencies to also have pinning inside that package. But that may not be always true in practice I guess.

I think ideally requirements.txt can be kept the same as setup.py's install_requires and then everything else goes into requirements-dev.txt. The CI tests should also run on an installed instance of the module instead of dev environment, so as to catch issues like https://github.com/octodns/octodns-azure/issues/7.

But I'm not sure of what's the industry standard practice. This sounds like a solved problem.

Reply:
We're way off in the weeds for this PR at this point, but it's fine :grin:.

> If I was pinning a package version, I'd generally expect its dependencies to also have pinning inside that package. But that may not be always true in practice I guess.

In general modules should never pin specific versions in their setup.py dependencies b/c that will require anything that uses them to have a hard requirement of that specific version. If there's a security vuln or major bug, anyone that uses the module can't access them. 

Another issue with it is if a project wants to use modules A and B which both require C. If A and B don't agree on a version then deps can't be satisfied. 

Modules should always require a `>=`, meaning this is the API/version I have expect to work as required by my module based on my testing and API/functionality at that point. Newer is fine, but not older.

> I think ideally requirements.txt can be kept the same as setup.py's install_requires and then everything else goes into requirements-dev.txt. The CI tests should also run on an installed instance of the module instead of dev environment, so as to catch issues like octodns/octodns-azure#7.

Full/exhaustive pinning should be done by the final project so that it has an expected and non-changing environment to work with. For modules that just means in dev & test (as all the octoDNS stuff does.) but for applications that should be for dev, test, and production. 

Pinning/freezing at that point ensures that things don't break unexpectedly when some module way down the chain updates. Those changes will only be picked up in PRs that explicitly update the requirements.txt.

> But I'm not sure of what's the industry standard practice. This sounds like a solved problem in general.

I think the above generally is the "best practice" give or take. In an ideal world the pinning also includes SHA verification of the packages. The pip ecosystem doesn't fully support that and relies on pypi to not let people change published packages. That only works so long as pypi is secure though...

Reply:
Makes sense, agreed totally not for this PR!